MaterialSkin/HTML/material/html/js/browse-functions.js

```diff
@@ -1882,6 +1882,39 @@ function browseHeaderAction(view, act, event, ignoreOpenMenus) {
         for (var i=0,len=sorts.length; i<len; ++i) {
             menuItems.push({key:sorts[i].key, label:sorts[i].label, selected:sort.by==sorts[i].key});
         }
+
+        if (LMS_VERSION>=90100 && ALBUM_SORTS_ACTION==act) {
```

Comment:
Calling this every time the menu shown is inefficient. The result of this call should be stored with the current item, and reused if possible.

Reply:
I was trying doing this, but as soon as I started pushing to an array other than `menuItems`, the application would no longer wait for the server command results. The problem is I don't understand how the async stuff works.

Reply:
It does not 'wait', the menuItems is 'reactive' - so vue.js will update UI when anything changes. What you need to do is copy the (parsed) details of this response somewhere, and **also** update menuIitems. If this cached version exists, then use that for menuItems.

Reply:
Ah, I understand, it's in the framework. No wonder I couldn't find it!

Reply:
OK, I'm about to push some changes which hopefully cache the roles command results correctly.

These changes require the LMS from my updated branch https://github.com/darrell-k/slimserver/tree/sorting-by-role

I've also decided to clear any role sort when changing the view, writing the default to local storage for new views (except if we're viewing all albums, because by definition all roles sorts would be available).

This avoids having a stored sort (in local storage) which might not apply to a new view, because the set of albums do not contain the previously-selected sort role.

There is commented out code for a different approach, which "pretends" the sort is the default 'album' one if the actually used sort is on a role which isn't used in a new album list. This works because the LMS query will return the albums_loop in album name sequence for a role sort where the role doesn't exist for any albums. This means if the user subsequently displays an album list where the previously used sort role exists, it would still be used. But this is a bit messy and possibly fragile.

I suppose the ideal solution would be to have `getAlbumSort` always validate a role sort against the current albums list, but I think that would mean really having to wait for the LMS rolesQuery to return.

Don't know what you think on this?

Anyway, lots more testing to do, but I thought I'd push this so you can see where I am right now. (And I know I have a few debug messages to remove!)